wrapper_streamsource.go

```diff
@@ -0,0 +1,167 @@
+
+	readCloser := &readerWrapper{
+		reader: r.stdin,
+		read: func(w *readerWrapper, bytes []byte) (n int, err error) {
+			n, err = w.reader.Read(bytes)
+			totalBytes += int64(n)
+			return
+		},
+		close: func(w *readerWrapper) error {
+			if totalBytes > 0 && r.stdin != nil {
+				r.stdin = nil
+				w.close = nil
+			}
+			return nil
+		},
+	}
+
+	return readCloser, nil
+}
+
+func (r *resticWrapper) prepareCommandStreamSource() (io.ReadCloser, error) {
+	clog.Debug("redirecting command output to the backup")
+	pipeReader, pipeWriter := io.Pipe()
+	bufferedWriter := bufio.NewWriterSize(pipeWriter, 8*1024)
+
+	commandSignals := make(chan os.Signal, 2)
+	signal.Notify(commandSignals, os.Interrupt, syscall.SIGTERM, syscall.SIGABRT)
+
+	go func() {
+		defer pipeWriter.Close()
+		defer bufferedWriter.Flush()
+		defer signal.Stop(commandSignals)
+
+		env := append(os.Environ(), r.getEnvironment()...)
+		env = append(env, r.getProfileEnvironment()...)
+
+		for i, sourceCommand := range r.profile.Backup.StdinCommand {
+			clog.Debugf("starting 'stdin-command' command %d/%d: %s", i+1, len(r.profile.Backup.StdinCommand), sourceCommand)
+			rCommand := newShellCommand(sourceCommand, nil, env, r.dryRun, commandSignals, nil)
+			rCommand.stdout = bufferedWriter
+			rCommand.stderr = term.GetErrorOutput()
+
+			_, stderr, err := runShellCommand(rCommand)
+			if err != nil {
+				// discard unflushed output
+				bufferedWriter.Reset(pipeWriter)
+				// push command error to reader
+				err = newCommandError(rCommand, stderr, fmt.Errorf("'stdin-command' on profile '%s': %w", r.profile.Name, err))
+				if closeError := pipeWriter.CloseWithError(err); closeError != nil {
+					clog.Errorf("Failed closing pipe for command '%s' after %w ; close error: %w", sourceCommand, err, closeError)
+				}
+				return
+			}
+		}
+	}()
+
+	closePipe := func() error {
+		defer func() {
+			clog.Debugf("stopping 'stdin-command'")
+			signal.Stop(commandSignals)
+			commandSignals <- os.Interrupt
+		}()
+		return pipeReader.Close()
+	}
+
+	// read from pipe to ensure the process started and returns content or error before restic is started
+	var initialReader io.Reader
+	{
+		initialBytes := make([]byte, 512)
+		if n, err := pipeReader.Read(initialBytes); err == nil || err == io.EOF {
+			clog.Debugf("initial %d bytes successfully read from 'stdin-command'", n)
+			initialReader = bytes.NewReader(initialBytes[:n])
+		} else {
+			_ = closePipe()
+			return nil, err
+		}
+	}
+
+	readCloser := &readerWrapper{
+		reader: io.MultiReader(initialReader, pipeReader),
+
+		read: func(w *readerWrapper, bytes []byte) (n int, err error) {
+			n, err = w.reader.Read(bytes)
+
+			// Stopping restic when stream source command fails while producing content
+			if err != nil && err != io.EOF {
+				clog.Errorf("interrupting '%s' after stdin read error: %s", r.command, err)
+				if runtime.GOOS == "windows" {
+					os.Exit(1) // no signalling on Windows, use os.Exit() instead
```

Comment:
This is brutal. I can't think of a better way though 😞

Reply:
Good news! a simple `return` does work on Windows: that closes stdin and stops restic gracefully.
Sorry it took me a while to test it as I found out Virtualbox is no longer working with Monterey so I had to dust off an old laptop 😄